File: mobile/androidApp/src/main/kotlin/dev/centraid/android/kit/Rows.kt
package dev.centraid.android.kit

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import centraid.screen.v1.Autosave
import centraid.screen.v1.ListRow
import centraid.screen.v1.SectionHead
import centraid.screen.v1.StatusChip
import dev.centraid.android.theme.centraidColor
import dev.centraid.android.theme.centraidType

/**
 * THE ONE ROW SHAPE (hand-off v17: "one row shape per app reused by every list
 * in it"). Title over meta, a trailing figure or word inked by [trailingTone],
 * its [chips] in draw order, the 2px hue rule at inline-start when [hueKey]
 * names a party hue, 44dp minimum.
 *
 * [a11y] replaces what the row says when set (a row is one stop, not four);
 * otherwise the visible words are merged. [pending] draws the trailing
 * `Saving` word — a write in flight on this row. [accessory] is the one slot
 * for per-row controls (Trash's Restore / Delete forever).
 */
@Composable
public fun CentraidRow(
    title: String,
    modifier: Modifier = Modifier,
    meta: String = "",
    trailing: String = "",
    trailingNote: String = "",
    trailingTone: StatusChip.Tone = StatusChip.Tone.TONE_UNSPECIFIED,
    chips: List<StatusChip> = emptyList(),
    hueKey: String = "",
    dimmed: Boolean = false,
    pending: Boolean = false,
    pendingLabel: String = KitWords.PENDING,
    a11y: String = "",
    testTag: String = "kit-row",
    onTap: (() -> Unit)? = null,
    accessory: (@Composable RowScope.() -> Unit)? = null,
) {
    val titleInk = if (dimmed) centraidColor("textFaint") else centraidColor("text")
    val softInk = if (dimmed) centraidColor("textFaint") else centraidColor("textSoft")
    Row(
        modifier
            .fillMaxWidth()
            .heightIn(min = KitGeometry.ROW_MIN)
            .then(if (onTap != null) Modifier.clickable(onClick = onTap) else Modifier)
            .padding(horizontal = KitGeometry.GUTTER, vertical = 8.dp)
            .testTag(testTag)
            .then(
                when {
                    accessory != null -> Modifier.semantics(mergeDescendants = false) { }
                    a11y.isNotEmpty() -> Modifier.clearAndSetSemantics {
                        contentDescription = a11y
                        if (onTap != null) role = Role.Button
                    }
                    else -> Modifier.semantics(mergeDescendants = true) {
                        if (onTap != null) role = Role.Button
                    }
                },
            ),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (hueKey.isNotEmpty()) {
            Box(
                Modifier
                    .padding(end = 10.dp)
                    .width(2.dp)
                    .height(28.dp)
                    .clip(RoundedCornerShape(1.dp))
                    .background(hueColor(hueKey)),
            )
        }
        Column(Modifier.weight(1f)) {
            Text(
                title,
                style = centraidType("body"),
                color = titleInk,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
            if (meta.isNotEmpty()) {
                Text(
                    meta,
                    style = centraidType("annotLabel"),
                    color = softInk,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
        chips.forEach { chip -> StatusChipView(chip, Modifier.padding(start = 8.dp)) }
        if (trailing.isNotEmpty() || pending) {
            // THE FIGURE: right-aligned, with its sub-label under it.
            Column(Modifier.padding(start = 8.dp), horizontalAlignment = Alignment.End) {
                if (trailing.isNotEmpty()) {
                    val figureInk = if (dimmed) titleInk else toneInk(trailingTone, neutral = "text")
                    Text(trailing, style = centraidType("bodyStrong"), color = figureInk, textAlign = TextAlign.End, maxLines = 1)
                }
                val note = if (pending) pendingLabel else trailingNote
                if (note.isNotEmpty()) {
                    Text(note, style = centraidType("annotLabel"), color = softInk, textAlign = TextAlign.End, maxLines = 1)
                }
            }
        }
        accessory?.invoke(this)
    }
}

/** A kit `ListRow`, drawn as the one row. */
@Composable
public fun CentraidRow(row: ListRow, modifier: Modifier = Modifier, onTap: (() -> Unit)? = null) {
    CentraidRow(
        title = row.title,
        modifier = modifier,
        meta = row.meta,
        trailing = row.trailing,
        trailingTone = row.trailing_tone,
        chips = row.chips,
        hueKey = row.hue_key,
        dimmed = row.dimmed,
        pending = row.pending,
        a11y = row.accessibility_label,
        testTag = "kit-row-${row.id}",
        onTap = onTap,
    )
}

/**
 * A CHIP: a finished label on a hairline in its tone's ink. `seam` is "not
 * yet", `net` is "owe / leaves", `warning` is the warning role; neutral is
 * soft ink. No fill — a chip is read, not pressed.
 */
@Composable
public fun StatusChipView(chip: StatusChip, modifier: Modifier = Modifier) {
    if (chip.label.isEmpty()) return
    val ink = toneInk(chip.tone, neutral = "textSoft")
    Text(
        chip.label,
        style = centraidType("annotLabel"),
        color = ink,
        maxLines = 1,
        modifier = modifier
            .border(KitGeometry.HAIRLINE, ink, RoundedCornerShape(4.dp))
            .padding(horizontal = 6.dp, vertical = 1.dp)
            .testTag("kit-chip"),
    )
}

/** A tone's ink role: `seam`, `net`, `warning`, or [neutral] for the rest. */
@Composable
public fun toneInk(tone: StatusChip.Tone, neutral: String): Color = centraidColor(
    when (tone) {
        StatusChip.Tone.TONE_SEAM -> "seam"
        StatusChip.Tone.TONE_NET -> "net"
        StatusChip.Tone.TONE_WARNING -> "warning"
        else -> neutral
    },
)

/** THE SECTION HEAD: a title, its count, and one text verb. */
@Composable
public fun SectionHeader(head: SectionHead, modifier: Modifier = Modifier, onVerb: (() -> Unit)? = null) {
    Row(
        modifier
            .fillMaxWidth()
            .padding(start = KitGeometry.GUTTER, end = 8.dp, top = 20.dp, bottom = 4.dp)
            .testTag("kit-section"),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        val count = head.count
        Text(
            if (count != null) "${head.title} · $count" else head.title,
            style = centraidType("smallStrong"),
            color = centraidColor("textSoft"),
            modifier = Modifier.weight(1f).semantics { heading() },
        )
        if (head.verb_label.isNotEmpty() && onVerb != null) {
            Text(
                head.verb_label,
                style = centraidType("annotLabelOn"),
                color = centraidColor("link"),
                modifier = Modifier
                    .heightIn(min = 36.dp)
                    .clip(RoundedCornerShape(KitGeometry.RADIUS))
                    .clickable(onClick = onVerb)
                    .padding(horizontal = 8.dp, vertical = 9.dp)
                    .semantics { role = Role.Button },
            )
        }
    }
}

/**
 * "SHOW MORE" at the foot of a paged list: drawn while there is a next page,
 * quiet while the page is on its way. The member pulls the next page; nothing
 * pages on its own.
 */
@Composable
public fun ShowMoreFooter(
    visible: Boolean,
    loading: Boolean,
    onMore: () -> Unit,
    modifier: Modifier = Modifier,
    label: String = KitWords.SHOW_MORE,
    loadingLabel: String = KitWords.LOADING_MORE,
) {
    if (!visible) return
    Box(modifier.fillMaxWidth().padding(KitGeometry.GUTTER), contentAlignment = Alignment.Center) {
        if (loading) {
            Text(loadingLabel, style = centraidType("annotLabel"), color = centraidColor("textFaint"))
        } else {
            QuietButton(label = label, testTag = "kit-show-more", onPress = onMore)
        }
    }
}

/** THE FIELD ROW: key, value, and a note under them. Read-only. */
@Composable
public fun FieldRow(key: String, value: String, modifier: Modifier = Modifier, note: String = "") {
    Column(
        modifier
            .fillMaxWidth()
            .heightIn(min = KitGeometry.ROW_MIN)
            .padding(horizontal = KitGeometry.GUTTER, vertical = 8.dp)
            .semantics(mergeDescendants = true) { },
    ) {
        Text(key, style = centraidType("eyebrow"), color = centraidColor("textSoft"))
        Text(value, style = centraidType("body"), color = centraidColor("text"))
        if (note.isNotEmpty()) Text(note, style = centraidType("annotLabel"), color = centraidColor("textFaint"))
    }
}

/**
 * EDITOR TEXT THAT CANNOT DROP A KEYSTROKE.
 *
 * The member's typing lives HERE, not in the machine's state: a field whose
 * value round-tripped through an event and a reduce would lose every character
 * typed while the last one was in flight. The machine's value replaces the
 * local text only when [reload] changes to a non-null value — the state's own
 * "this came from the vault" signal (for an autosave editor: the baseline,
 * while the phase is CLEAN) — so a remote load lands and typing never fights it.
 */
@Composable
public fun rememberEditorText(value: String, reload: Any?): MutableState<String> {
    val text = remember { mutableStateOf(value) }
    LaunchedEffect(reload) {
        if (reload != null && text.value != value) text.value = value
    }
    return text
}

/**
 * AN EDITABLE FIELD ROW: key over a bare text field. [onEdit] is sent only for
 * a difference from [value], from local text held by [rememberEditorText].
 */
@Composable
public fun EditableFieldRow(
    key: String,
    value: String,
    reload: Any?,
    onEdit: (String) -> Unit,
    modifier: Modifier = Modifier,
    placeholder: String = "",
    singleLine: Boolean = true,
    style: String = "body",
    testTag: String = "kit-field",
) {
    val text = rememberEditorText(value, reload)
    Column(modifier.fillMaxWidth().padding(horizontal = KitGeometry.GUTTER, vertical = 6.dp)) {
        if (key.isNotEmpty()) Text(key, style = centraidType("eyebrow"), color = centraidColor("textSoft"))
        Box(Modifier.fillMaxWidth().heightIn(min = 36.dp), contentAlignment = Alignment.CenterStart) {
            if (text.value.isEmpty() && placeholder.isNotEmpty()) {
                Text(
                    placeholder,
                    style = centraidType(style),
                    color = centraidColor("textFaint"),
                    modifier = Modifier.clearAndSetSemantics { },
                )
            }
            BasicTextField(
                value = text.value,
                onValueChange = { next ->
                    text.value = next
                    if (next != value) onEdit(next)
                },
                singleLine = singleLine,
                textStyle = centraidType(style).copy(color = centraidColor("text")),
                cursorBrush = SolidColor(centraidColor("text")),
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag(testTag)
                    .semantics { contentDescription = key.ifEmpty { placeholder } },
            )
        }
    }
}

/** A CHOICE FIELD ROW: key and the current value; a tap opens the choice sheet. */
@Composable
public fun ChoiceFieldRow(
    key: String,
    value: String,
    onTap: () -> Unit,
    modifier: Modifier = Modifier,
    testTag: String = "kit-choice",
) {
    Row(
        modifier
            .fillMaxWidth()
            .heightIn(min = KitGeometry.ROW_MIN)
            .clickable(onClick = onTap)
            .padding(horizontal = KitGeometry.GUTTER, vertical = 8.dp)
            .testTag(testTag)
            .clearAndSetSemantics {
                contentDescription = "$key, $value"
                role = Role.Button
            },
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(key, style = centraidType("body"), color = centraidColor("textSoft"), modifier = Modifier.weight(1f))
        Text(value, style = centraidType("body"), color = centraidColor("text"))
        CentraidIcon(
            iconKey = "ChevronRight",
            tint = centraidColor("textFaint"),
            size = 16.dp,
            modifier = Modifier.padding(start = 6.dp),
        )
    }
}

/** What an autosave says, in words: the status line an [EditorRoom] hosts. */
public fun autosaveLabel(autosave: Autosave?): String {
    if (autosave == null) return ""
    return when (autosave.phase) {
        Autosave.Phase.PHASE_DIRTY, Autosave.Phase.PHASE_SAVING -> KitWords.SAVING
        Autosave.Phase.PHASE_REFUSED -> autosave.failure?.sentence?.ifEmpty { null } ?: KitWords.NOT_SAVED
        Autosave.Phase.PHASE_CONFLICT -> KitWords.CHANGED_ELSEWHERE
        Autosave.Phase.PHASE_CLEAN, Autosave.Phase.PHASE_SAVED -> KitWords.SAVED
        Autosave.Phase.PHASE_UNSPECIFIED -> ""
    }
}

/**
 * THE AUTOSAVE STATUS LINE. A refusal is `net` ink and its sentence, over the
 * words and never instead of them; everything else is quiet.
 */
@Composable
public fun AutosaveStatus(autosave: Autosave?, modifier: Modifier = Modifier) {
    val label = autosaveLabel(autosave)
    if (label.isEmpty()) return
    val refused = autosave?.phase == Autosave.Phase.PHASE_REFUSED
    Text(
        label,
        style = centraidType("annotLabel"),
        color = if (refused) centraidColor("net") else centraidColor("textSoft"),
        maxLines = 2,
        modifier = modifier.testTag("kit-autosave"),
    )
}

/**
 * A party hue to its colour. The state carries either the theme's colour role
 * itself (`cRose` — People) or a wheel key (`rose` — Agenda, Tally), and both
 * draw the same role. An unknown value draws slate, the neutral for "no party".
 */
@Composable
public fun hueColor(key: String): Color {
    val role = when {
        key in PARTY_ROLES -> key
        key in PARTY_HUES -> "c" + key.replaceFirstChar { it.uppercase() }
        else -> "cSlate"
    }
    return centraidColor(role)
}

/** `packages/design`'s eight party hues. */
private val PARTY_HUES: Set<String> =
    setOf("rose", "amber", "ochre", "forest", "teal", "slate", "indigo", "violet")

/** The same eight, as the theme's colour roles. */
private val PARTY_ROLES: Set<String> = PARTY_HUES.mapTo(mutableSetOf()) { "c" + it.replaceFirstChar(Char::uppercaseChar) }

File: mobile/iosApp/Sources/Kit/Rows.swift
import SwiftUI

// ONE ROW SHAPE, ONE SECTION HEAD, ONE FIELD ROW (K5; handoff README l.22-27:
// "one row shape per app reused by every list in it; the section head with a
// count and a text verb; the field row of key, value and note; the chip").

/// THE LIST ROW: title over meta, a trailing figure, its chips in draw order,
/// a 2pt hue rule at the inline start, 44pt minimum.
///
/// `hueKey` is a colour role (`cTeal`, `net`, …) — empty draws no rule.
/// `dimmed` recedes the ink (a settled expense, a done task); `pending` says a
/// write for this row is in flight, which is the screen's knowledge and never
/// a column in the vault.
struct CentraidRow: View {
    let title: String
    var meta: String = ""
    var trailing: String = ""
    var trailingTone: String = "text"
    var chips: [Centraid_Screen_V1_StatusChip] = []
    var hueKey: String = ""
    var dimmed: Bool = false
    var pending: Bool = false
    var pendingLabel: String = "Saving"
    var accessibility: String = ""
    var identifier: String = ""
    var onTap: (() -> Void)? = nil

    @Environment(\.colorScheme) private var scheme

    /// A row the machine already folded into the kit's `ListRow`.
    init(_ row: Centraid_Screen_V1_ListRow, onTap: (() -> Void)? = nil) {
        self.init(
            title: row.title,
            meta: row.meta,
            trailing: row.trailing,
            trailingTone: StatusChipView.ink(row.trailingTone, neutral: "text"),
            chips: row.chips,
            hueKey: row.hueKey,
            dimmed: row.dimmed,
            pending: row.pending,
            accessibility: row.accessibilityLabel,
            identifier: row.id.isEmpty ? "" : "row-\(row.id)",
            onTap: onTap
        )
    }

    init(
        title: String,
        meta: String = "",
        trailing: String = "",
        trailingTone: String = "text",
        chip: Centraid_Screen_V1_StatusChip? = nil,
        chips: [Centraid_Screen_V1_StatusChip] = [],
        hueKey: String = "",
        dimmed: Bool = false,
        pending: Bool = false,
        pendingLabel: String = "Saving",
        accessibility: String = "",
        identifier: String = "",
        onTap: (() -> Void)? = nil
    ) {
        self.title = title
        self.meta = meta
        self.trailing = trailing
        self.trailingTone = trailingTone
        self.chips = (chip.map { [$0] } ?? []) + chips
        self.hueKey = hueKey
        self.dimmed = dimmed
        self.pending = pending
        self.pendingLabel = pendingLabel
        self.accessibility = accessibility
        self.identifier = identifier
        self.onTap = onTap
    }

    var body: some View {
        Group {
            if let onTap {
                Button(action: onTap) { row }.buttonStyle(KitRowPress())
            } else {
                row
            }
        }
        .accessibilityElement(children: .combine)
        .modifier(OptionalLabel(label: accessibility))
        .modifier(OptionalIdentifier(identifier: identifier))
    }

    private var row: some View {
        HStack(spacing: 12) {
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .centraidType("body")
                    .foregroundStyle(Theme.color(dimmed ? "textFaint" : "text", scheme))
                    .lineLimit(2)
                if !meta.isEmpty || pending {
                    Text(pending ? (meta.isEmpty ? pendingLabel : "\(meta) · \(pendingLabel)") : meta)
                        .centraidType("annotLabel")
                        .foregroundStyle(Theme.color(dimmed ? "textFaint" : "textSoft", scheme))
                        .lineLimit(1)
                }
            }
            Spacer(minLength: 8)
            ForEach(Array(chips.enumerated()), id: \.offset) { _, chip in StatusChipView(chip) }
            if !trailing.isEmpty {
                Text(trailing)
                    .centraidType("smallStrong")
                    .monospacedDigit()
                    .foregroundStyle(Theme.color(dimmed ? "textFaint" : trailingTone, scheme))
                    .lineLimit(1)
            }
        }
        .padding(.horizontal, CentraidGeometry.pageMargin)
        .padding(.vertical, 8)
        .frame(minHeight: CentraidGeometry.targetMinCoarse)
        .frame(maxWidth: .infinity, alignment: .leading)
        .overlay(alignment: .leading) {
            if let hue = KitHue.role(hueKey) {
                Rectangle()
                    .fill(Theme.color(hue, scheme))
                    .frame(width: 2)
                    .padding(.vertical, 8)
            }
        }
        .overlay(alignment: .bottom) {
            Rectangle()
                .fill(Theme.color("line", scheme))
                .frame(height: CentraidGeometry.hairline)
                .padding(.leading, CentraidGeometry.pageMargin)
        }
        .contentShape(Rectangle())
    }
}

/// The press every tappable kit row shows: `bgPress` under the row while held.
struct KitRowPress: ButtonStyle {
    @Environment(\.colorScheme) private var scheme

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .background(configuration.isPressed ? Theme.color("bgPress", scheme) : Color.clear)
    }
}

private struct OptionalLabel: ViewModifier {
    let label: String
    func body(content: Content) -> some View {
        if label.isEmpty { content } else { content.accessibilityLabel(label) }
    }
}

private struct OptionalIdentifier: ViewModifier {
    let identifier: String
    func body(content: Content) -> some View {
        if identifier.isEmpty { content } else { content.accessibilityIdentifier(identifier) }
    }
}

/// A STATUS OR A SET CHOICE, AS A CHIP: a finished label in a hairline capsule
/// whose ink is the tone. `seam` is "not yet", `net` is "owe / leaves",
/// `warning` is a caution; everything else is soft ink.
struct StatusChipView: View {
    let chip: Centraid_Screen_V1_StatusChip

    @Environment(\.colorScheme) private var scheme

    init(_ chip: Centraid_Screen_V1_StatusChip) { self.chip = chip }

    private var tone: String { Self.ink(chip.tone, neutral: "textSoft") }

    /// A tone's ink role. `neutral` is what NEUTRAL and UNSPECIFIED draw — soft
    /// ink on a chip, plain ink on a row's trailing figure.
    static func ink(_ tone: Centraid_Screen_V1_StatusChip.Tone, neutral: String) -> String {
        switch tone {
        case .seam: return "seam"
        case .net: return "net"
        case .warning: return "warning"
        case .neutral, .unspecified, .UNRECOGNIZED: return neutral
        }
    }

    var body: some View {
        Text(chip.label)
            .centraidType("annotLabel")
            .foregroundStyle(Theme.color(tone, scheme))
            .lineLimit(1)
            .padding(.horizontal, 8)
            .padding(.vertical, 2)
            .overlay(
                Capsule().strokeBorder(Theme.color(tone == "textSoft" ? "line" : tone, scheme), lineWidth: CentraidGeometry.hairline)
            )
    }
}

/// THE SECTION HEAD: a title, its count, and a text verb at the end ("See
/// all", "Clear"). Never a chevron-only control — the verb is a word.
struct SectionHeader: View {
    let head: Centraid_Screen_V1_SectionHead
    var onVerb: (() -> Void)? = nil

    @Environment(\.colorScheme) private var scheme

    init(_ head: Centraid_Screen_V1_SectionHead, onVerb: (() -> Void)? = nil) {
        self.head = head
        self.onVerb = onVerb
    }

    init(title: String, count: UInt32? = nil, verb: String = "", onVerb: (() -> Void)? = nil) {
        var head = Centraid_Screen_V1_SectionHead()
        head.title = title
        if let count { head.count = count }
        head.verbLabel = verb
        self.init(head, onVerb: onVerb)
    }

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: 6) {
            Text(head.title)
                .centraidType("eyebrow")
                .foregroundStyle(Theme.color("textSoft", scheme))
                .accessibilityAddTraits(.isHeader)
            if head.hasCount {
                Text("\(head.count)")
                    .centraidType("eyebrow")
                    .foregroundStyle(Theme.color("textFaint", scheme))
            }
            Spacer(minLength: 0)
            if !head.verbLabel.isEmpty, let onVerb {
                Button(action: onVerb) {
                    Text(head.verbLabel)
                        .centraidType("annotLabelOn")
                        .foregroundStyle(Theme.color("link", scheme))
                        .frame(minHeight: CentraidGeometry.targetMinFine)
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
            }
        }
        .padding(.horizontal, CentraidGeometry.pageMargin)
        .padding(.top, 16)
        .padding(.bottom, 4)
    }
}

/// THE NEXT PAGE, ASKED FOR. A paged list's foot: shown only while the data
/// names a next cursor, and quiet while a page-one re-read is in flight (the
/// kit's `PagedList` refuses a next page then anyway).
struct ShowMoreFooter: View {
    let visible: Bool
    var loading: Bool = false
    var label: String = "Show more"
    let onMore: () -> Void

    var body: some View {
        if visible {
            HStack {
                Spacer()
                KitOutlineButton(label: label, action: onMore)
                    .disabled(loading)
                    .opacity(loading ? 0.5 : 1)
                    .accessibilityIdentifier("kit-show-more")
                Spacer()
            }
            .padding(.vertical, 12)
        }
    }
}

/// THE FIELD ROW: key over value, with an optional note under — a detail
/// page's read-only fact.
struct FieldRow: View {
    let key: String
    let value: String
    var note: String = ""

    @Environment(\.colorScheme) private var scheme

    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(key)
                .centraidType("annotLabel")
                .foregroundStyle(Theme.color("textSoft", scheme))
            Text(value)
                .centraidType("body")
                .foregroundStyle(Theme.color("text", scheme))
            if !note.isEmpty {
                Text(note)
                    .centraidType("annotLabel")
                    .foregroundStyle(Theme.color("textFaint", scheme))
            }
        }
        .padding(.horizontal, CentraidGeometry.pageMargin)
        .padding(.vertical, 8)
        .frame(maxWidth: .infinity, minHeight: CentraidGeometry.targetMinCoarse, alignment: .leading)
        .overlay(alignment: .bottom) { KitHairline() }
        .accessibilityElement(children: .combine)
    }
}

/// A FIELD THE MEMBER TYPES INTO, inside an editor. The text is the view's own
/// `@State` (the only place a keystroke exists before the machine hears it);
/// `onEdit` hands each difference over as the screen's edit event.
struct EditableFieldRow: View {
    let key: String
    @Binding var text: String
    var placeholder: String = ""
    var identifier: String = ""
    let onEdit: (String) -> Void

    @Environment(\.colorScheme) private var scheme

    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(key)
                .centraidType("annotLabel")
                .foregroundStyle(Theme.color("textSoft", scheme))
            TextField(placeholder, text: $text)
                .centraidType("body")
                .foregroundStyle(Theme.color("text", scheme))
                .onChange(of: text) { _, typed in onEdit(typed) }
                .accessibilityLabel(key)
                .modifier(OptionalIdentifier(identifier: identifier))
        }
        .padding(.horizontal, CentraidGeometry.pageMargin)
        .padding(.vertical, 8)
        .frame(maxWidth: .infinity, minHeight: CentraidGeometry.targetMinCoarse, alignment: .leading)
        .overlay(alignment: .bottom) { KitHairline() }
    }
}

/// A FIELD WHOSE VALUE IS A CHOICE: tapping opens the choice (a sheet — a
/// choice never pushes, #1015 D4).
struct ChoiceFieldRow: View {
    let key: String
    let value: String
    var identifier: String = ""
    let onTap: () -> Void

    @Environment(\.colorScheme) private var scheme

    var body: some View {
        Button(action: onTap) {
            HStack {
                Text(key)
                    .centraidType("body")
                    .foregroundStyle(Theme.color("text", scheme))
                Spacer(minLength: 8)
                Text(value)
                    .centraidType("body")
                    .foregroundStyle(Theme.color("textSoft", scheme))
                    .lineLimit(1)
                CentraidIconView(iconKey: "ChevronRight", tint: Theme.color("textFaint", scheme), size: 14)
            }
            .padding(.horizontal, CentraidGeometry.pageMargin)
            .frame(maxWidth: .infinity, minHeight: CentraidGeometry.targetMinCoarse)
            .overlay(alignment: .bottom) { KitHairline() }
            .contentShape(Rectangle())
        }
        .buttonStyle(KitRowPress())
        .accessibilityElement(children: .combine)
        .modifier(OptionalIdentifier(identifier: identifier))
    }
}

/// The row divider every kit row draws: a hairline inset by the page margin.
struct KitHairline: View {
    @Environment(\.colorScheme) private var scheme

    var body: some View {
        Rectangle()
            .fill(Theme.color("line", scheme))
            .frame(height: CentraidGeometry.hairline)
            .padding(.leading, CentraidGeometry.pageMargin)
    }
}

/// A ROW'S HUE, AS THE THEME SPELLS IT. Machines send the colour role
/// (`cRose`); a few still send the wheel key (`rose`). Both resolve here, and
/// anything the emitted table lacks draws no rule — `Theme.color` traps on an
/// unknown role, and a row must never crash on what the vault stored.
enum KitHue {
    static func role(_ key: String) -> String? {
        guard !key.isEmpty else { return nil }
        if centraidColorRoles.contains(key) { return key }
        let wheel = "c" + key.prefix(1).uppercased() + key.dropFirst()
        return centraidColorRoles.contains(wheel) ? wheel : nil
    }
}
